R/PACKAGES.R

```diff
@@ -1,58 +1,158 @@
-#' @param type Package type: `"source"`, `"mac.binary"`, or `"win.binary"`.
-#' @param contrib_url URL of the contrib sections of the repository.
-#' @param quiet Passed on to `curl::curl_download()`.
+#' @param base_url Base URL of the repository. Can be a local dir path.
+#' @param platform Platform name.
+#' @param r_version R version.
 #'
-#' @returns Path of downloaded file (invisibly).
+#' @return data.frame
 #'
 #' @export
-get_PACKAGES <- function(destpath,
-                         filename = "PACKAGES",
-                         repos = getOption("repos"),
-                         type = getOption("pkgType"),
-                         contrib_url = utils::contrib.url(repos, type),
-                         quiet = FALSE) {
-  # TODO: argument checking.
-  url <- file.path(contrib_url, "PACKAGES")
-  destfile <- file.path(destpath, filename)
-  fs::dir_create(destpath)
-  curl::curl_download(url = url, destfile = destfile, quiet = quiet)
-}
+get_packages <- function(
+    base_url = .config$remote_base,
+    platform = .config$platform,
+    r_version = .config$r_version) {
+  platform <- match.arg(platform, RHUB_REPO_PLATFORMS)
 
-#' @export
-path_to_PACKAGES <- function(path) {
-  fs::dir_ls(path, recurse = TRUE, glob = "*/PACKAGES")
+  full_path <- file.path(base_url, platform, r_version) |>
+    utils::contrib.url(type = .get_repos_type(platform)) |>
+    file.path("PACKAGES")
+
+  connection <- .get_connection(full_path)
+
+  on.exit(close(connection))
+
+  read.dcf(connection, all = TRUE) |>
+    as.data.frame()
 }
 
+#' Identify new or newer packages.
+#'
+#' New packages - are the packages that are present in the remote repository
+#' but not in the local repository.
+#'
+#' Newer packages - are the packages that are present in both repositories,
+#' but the version in the remote repository is newer.
+#'
+#' @param remote_packages Data frame with remote packages.
+#' @param local_packages Data frame with local packages.
+#'
+#' @return data.frame
+#'
 #' @export
-path_to_RISKFIELDS <- function(path) {
-  fs::dir_ls(path, recurse = TRUE, glob = "*/RISKFIELDS")
+diff_packages <- function(remote_packages, local_packages) {
+  stopifnot(
+    "Package" %in% names(remote_packages),
+    "Package" %in% names(local_packages),
+    "Version" %in% names(remote_packages),
+    "Version" %in% names(local_packages)
+  )
+
+  combined_packages <- merge(
+    x = remote_packages,
+    y = local_packages,
+    all.x = TRUE,
+    by = "Package",
+    suffixes = c("_remote", "_local")
+  )
+
+  is_new_or_newer <- mapply(
+    utils::compareVersion,
+    combined_packages$Version_remote,
+    combined_packages$Version_local
+  )
+
+  combined_packages$VersionDifference <- ifelse(
+    is_new_or_newer > 0,
+    paste(combined_packages$Version_remote, ">", combined_packages$Version_local),
+    combined_packages$Version_local
+  )
+
+  combined_packages[is_new_or_newer > 0, c("Package", "VersionDifference")]
```

Comment:
You are right! My idea was to return some information about the version differences so that it could be printed or reported somehow. I ended up not reporting anything in the end, so I'd rather refactor it to only return package names 👍